silvestrst-crypto/WalletConnectRust
# feat: create a simple Sign API example on top of the WS client

# Description

This change adds Sign API framework, and a simple example client to demonstrate pairing and session creation capabilities.

This change introduces the following functionality:
- Sign API payload serializable/deserializable data types
- Payload encryption and decryption
- Session Diffie–Hellman key derivation
- Pairing URI parsing
- Simple WS based client

Resolves #47

## How Has This Been Tested?

- [x] Use example client to pair and establish the session with an example (DApp)[https://react-app.walletconnect.com/]:
      - Click on "Goerli" and then "connect"
      - In a new pop-up, click "New Pairing"
      - Copy the Pairing URI
      - In the terminal, `cd path/to/WalletConnectRust`
      - `.../WalletConnectRust$ cargo run --example session "<copied URI>"`
      - DApp should now display the session window
      - In the DApp click disconnect

![image](https://github.com/silvestrst-crypto/WalletConnectRust/assets/103181555/24d1bc4c-e11c-4723-be41-00c9c62788be)



## Due Diligence


## TODO

- [ ] Typed errors instead of unnamed anyhow errors
- [ ] More unittests
- [ ] Better documentation, code comments and fuction descriptions
- [x] `crypto/session.rs` refactor

## Additional information

WCv2 uses layered approach. The relay protocol messages (currently only IRN, but there might be other in future), are used to carry the useful payload. In terms of this change the useful payload is Sign API messages.

Useful materials:
https://specs.walletconnect.com/2.0/specs/clients/core/pairing
https://specs.walletconnect.com/2.0/specs/clients/sign/session-proposal

https://specs.walletconnect.com/2.0/specs/servers/relay/relay-server-rpc
https://specs.walletconnect.com/2.0/specs/clients/sign/rpc-methods
